File: python/external-user-service/OrderService/cache/in_memory_jwks_cache.py
"""
In-Memory JWKS Key Cache Implementation
Thread-safe in-memory caching for JWT signing keys
"""

import asyncio
from datetime import datetime, timedelta
from typing import Optional, Dict
from concurrent.futures import ThreadPoolExecutor
import logging
import base64
from cryptography.hazmat.primitives.asymmetric.rsa import RSAPublicNumbers
from cryptography.hazmat.backends import default_backend

from .jwks_key_cache import IJwksKeyCache, CachedSecurityKey, JwksKey

logger = logging.getLogger(__name__)


class CacheEntry:
    """Internal cache entry with expiration"""
    
    def __init__(self, jwk: JwksKey, expires_at: datetime):
        self.jwk = jwk
        self.expires_at = expires_at


class InMemoryJwksKeyCache(IJwksKeyCache):
    """In-memory implementation of JWKS key cache"""
    
    def __init__(self):
        self._cache: Dict[str, CacheEntry] = {}
        self._lock = asyncio.Lock()
    
    async def try_get_async(self, kid: str) -> Optional[CachedSecurityKey]:
        """Try to get a cached security key by kid"""
        async with self._lock:
            entry = self._cache.get(kid)
            
            if entry:
                if datetime.utcnow() <= entry.expires_at:
                    # Create security key from cached JWKS key
                    try:
                        public_key = self._create_security_key(entry.jwk)
                        logger.debug(f"Using cached key for kid: {kid}")
                        return CachedSecurityKey(public_key, datetime.utcnow())
                    except Exception as ex:
                        logger.error(f"Error creating security key from cached JWKS: {ex}")
                        # Remove invalid entry
                        del self._cache[kid]
                else:
                    # Entry expired, remove it
                    del self._cache[kid]
                    logger.debug(f"Cached key expired for kid: {kid}")
            
            return None
    
    async def set_async(self, kid: str, jwk: JwksKey, ttl: timedelta):
        """Set a JWKS key in cache with TTL"""
        async with self._lock:
            expires_at = datetime.utcnow() + ttl
            self._cache[kid] = CacheEntry(jwk, expires_at)
            logger.debug(f"Cached key with kid: {kid}, expires at: {expires_at}")
    
    async def remove_async(self, kid: str):
        """Remove a key from cache"""
        async with self._lock:
            if kid in self._cache:
                del self._cache[kid]
                logger.debug(f"Removed cached key for kid: {kid}")
    
    def _create_security_key(self, jwks_key: JwksKey):
        """Create RSA public key from JWKS key (equivalent to .NET CreateSecurityKeyStatic)"""
        try:
            # Decode base64url encoded values
            n_bytes = self._base64url_decode(jwks_key.n)
            e_bytes = self._base64url_decode(jwks_key.e)
            
            # Convert to integers
            n = int.from_bytes(n_bytes, byteorder="big")
            e = int.from_bytes(e_bytes, byteorder="big")
            
            # Create RSA public key
            public_key = RSAPublicNumbers(e, n).public_key(default_backend())
            
            return public_key
            
        except Exception as ex:
            logger.error(f"Error creating RSA public key from JWKS: {ex}")
            raise
    
    def _base64url_decode(self, input_str: str) -> bytes:
        """Decode base64url string"""
        padding = 4 - len(input_str) % 4
        if padding != 4:
            input_str += '=' * padding
        return base64.urlsafe_b64decode(input_str)

File: nodejs/external-user-service/OrderService/src/cache/in.memory.jwks.cache.ts
/**
 * In-Memory JWKS Key Cache
 * TypeScript equivalent to Python in-memory JWKS cache
 */

import { IJwksKeyCache, JwksKey } from './jwks.key.cache.interface';
import { logger } from '../common/logger';

interface CachedKey {
    key: JwksKey;
    expiresAt: Date;
}

export class InMemoryJwksKeyCache implements IJwksKeyCache {
    private cache: Map<string, CachedKey> = new Map();

    async tryGetAsync(kid: string): Promise<JwksKey | null> {
        try {
            const cached = this.cache.get(kid);
            
            if (!cached) {
                logger.debug(`Cache miss for key ID: ${kid}`);
                return null;
            }
            
            // Check if expired
            if (cached.expiresAt <= new Date()) {
                logger.debug(`Cache expired for key ID: ${kid}`);
                this.cache.delete(kid);
                return null;
            }
            
            logger.debug(`Cache hit for key ID: ${kid}`);
            return cached.key;
        } catch (error) {
            logger.error(`Error getting key from in-memory cache: ${error}`);
            return null;
        }
    }

    async setAsync(kid: string, jwk: JwksKey, ttlMinutes: number): Promise<void> {
        try {
            const expiresAt = new Date();
            expiresAt.setMinutes(expiresAt.getMinutes() + ttlMinutes);
            
            this.cache.set(kid, {
                key: jwk,
                expiresAt
            });
            
            logger.debug(`Key cached in-memory for ID: ${kid}, expires at: ${expiresAt.toISOString()}`);
        } catch (error) {
            logger.error(`Error setting key in in-memory cache: ${error}`);
        }
    }

    async removeAsync(kid: string): Promise<void> {
        try {
            const deleted = this.cache.delete(kid);
            if (deleted) {
                logger.debug(`Key removed from in-memory cache: ${kid}`);
            }
        } catch (error) {
            logger.error(`Error removing key from in-memory cache: ${error}`);
        }
    }

    async clearAsync(): Promise<void> {
        try {
            this.cache.clear();
            logger.debug('In-memory cache cleared');
        } catch (error) {
            logger.error(`Error clearing in-memory cache: ${error}`);
        }
    }

    // Cleanup expired keys periodically
    public startCleanupTimer(): void {
        setInterval(() => {
            this.cleanupExpiredKeys();
        }, 5 * 60 * 1000); // Every 5 minutes
    }

    private cleanupExpiredKeys(): void {
        const now = new Date();
        let cleanedCount = 0;
        
        for (const [kid, cached] of this.cache.entries()) {
            if (cached.expiresAt <= now) {
                this.cache.delete(kid);
                cleanedCount++;
            }
        }
        
        if (cleanedCount > 0) {
            logger.debug(`Cleaned up ${cleanedCount} expired keys from in-memory cache`);
        }
    }
}
